### PR TITLE
Aho corasick

### LINKED ISSUE #931: [Feature Request] <Aho-Corasick >
## 🚀 Feature Request: Aho-Corasick
Is your feature request related to a problem? Please describe.
Our current approach for matching multiple patterns in a text is slow and cumbersome, leading to performance issues. The Aho-Corasick algorithm would resolve these problems.

Describe the solution you'd like
Implement the Aho-Corasick algorithm to allow for efficient multi-pattern matching using a trie structure, improving search times significantly.

Describe alternatives you've considered
I considered using simpler algorithms but realized they cannot match multiple patterns simultaneously as efficiently as Aho-Corasick.

Additional context
Adding this algorithm will greatly enhance the application's ability to handle complex string searching tasks, especially in applications like search engines and data processing.

**Would you like to work on this feature?**  
- [x] Yes  
- [ ] No

### PR DESCRIPTION
**Pull request by priyashuu**
### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #931

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/String algorithm/AHO-corasick.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+- **Text Search:** Useful in applications such as search engines, text editors, and data mining where multiple patterns need to be searched.
+- **Intrusion Detection Systems:** Often used to detect multiple predefined patterns in network traffic.
+- **DNA Sequence Analysis:** Helps in finding multiple patterns within genetic sequences.
+
+## Conclusion
+
+The **Aho-Corasick Algorithm** is an optimal solution for problems involving multiple pattern matching. Its combination of a trie and failure links ensures that it can handle large datasets and multiple queries efficiently.
+```
+
+---
```
Comment (ajay-dhangar) [inline docs/String algorithm/AHO-corasick.md]: delete line 256 to 258  we not need this

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update code based on suggestions

File: docs/String algorithm/AHO-corasick.md (hunk 1 above)
Comment (priyashuu) [reply docs/String algorithm/AHO-corasick.md]: done

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**